docs/document-scope.md

```diff
@@ -0,0 +1,73 @@
+- `sharedModelManager`
+- `appConfig`
+
+I'll call these "services".
+
+The environment object of a MST tree has to be provided when the root of the tree is created and this object instance can't be changed. This object is not observable. And according to the docs the top level properties of this object should not be modified. I think this is because MST does try to merge environments when a node with an environment is added to another tree. However nested environments are not supported like React supports nested context's.
+
+### Cases
+- `getSharedModelManager`
+- `getAppConfig`
+
+### Benefits
+Any MST tree can have an environment that provides the necessary services.
+
+### Downsides
+The construction of this environment is kind of awkward because it often requires a circular reference. In many cases the services need access to the tree itself. So the tree refers to the service and the service refers to the tree.
+
+So the environment service has to be created, then the root node created with the environment service, and finally the root node needs to set back on the environment service. See `createDocumentModel` for an example of this.
+
+The top level properties of the environment object are not supposed to be modified after it is created, based on this "shallowly immutable" note here: https://mobx-state-tree.js.org/concepts/dependency-injection
+However we are doing this when the appConfig is added to the environment object in `Documents#add`
+
+# View layer
+
+## React Context
+
+### Cases
+- `AddTilesContext` provided by the Canvas component. `DataSetViewButton` uses this add new tiles to the document.
+- `TileApiInterfaceContext` provided by the Canvas component. This is used internally for the tileApi mechanism described below
+- `DocumentContextReact` provided by `EditableDocumentContent`, `CollapsibleDocumentsSection`, and `DocumentCollectionByType`. This provides basic info about the document (type, key, title, originDoc) and methods for setting and getting properties on the document. This is accessed by `useImageContentUrl`, but it does not seem to actually be using it. It is also accessed by `DataflowProgram` but again doesn't seem to be used.
+- `EditableTileApiInterfaceRefContext` provided by the `EditableDocumentContent`. It is a simple React ref object so basically just an object with `current`. The `Canvas` sets this `current` to be the same `tileApiInterface` that is available via the `TileApiInterfaceContext`. I'm not sure why this is. Its existence is checked by the `Toolbar`'s `getUniqueTitle`. I'm not sure what this existence check is for. Its `deleteSelection` function is used by the `Toolbar`'s `handleDelete`.
+
+These next contexts are not at the document level but it seemed good to include them for completeness:
+- `TileModelContext` this is provided by the `TileComponent`. Its value is the `TileModel` instance of the tile.
```

Comment:
FYI: this kind of context has presented undo/redo issues in CODAP v3. If a tile is deleted, any references to the tile model are invalid until the next React render cycle updates the context. If there are MobX reactions or `useCallback()`s that refer to the model they can trigger before the context is updated, leading to MST warnings.

Reply:
One solution for the reactions is you can call `addDisposer(model, reactionDisposer)`.  This way they can be cleaned up right away. But the `useCallback()` and even in some cases just a simple access in the render method can trigger these problems. The only way I know to handle that is to add alive checks before accessing the model. Perhaps we can devise a system where we clean up the tile component first before actually deleting the tile from MST. This would require a two phase delete which might introduce more problems than just adding the alive checks.